boards: arm: Add CONFIG_PERSISTENT_GECKO_RTCC to efm32pg_stk3402a

DRuffer-tmo opened the pull request:
Add us to the CODEOWNERS of the efm32pg_stk3402a.
Add a Kconfig for turning on the COUNTER_GECKO_RTCC. Add counter to the twister yaml supported devices. Change the rtcc0 prescaler from 1 to 32768.
Set multiple CONFIG settings to maintain the clock. Add PERSISTENT_GECKO_RTCC to support this feature. Add list of SiLabs devices where this feature works. Set the RTCC manually in the shell's date set command.

DRuffer-tmo committed:
boards: arm: Add CONFIG_PERSISTENT_GECKO_RTCC to efm32pg_stk3402a

Add us to the CODEOWNERS of the efm32pg_stk3402a.
Add a Kconfig for turning on the COUNTER_GECKO_RTCC.
Add counter to the twister yaml supported devices.
Change the rtcc0 prescaler from 1 to 32768.
Set multiple CONFIG settings to maintain the clock.
Add PERSISTENT_GECKO_RTCC to support this feature.
Add list of SiLabs devices where this feature works.
Set the RTCC manually in the shell's date set command.

Signed-off-by: Dennis Ruffer <Dennis.Ruffer1@T-Mobile.com>

fkokosinski commented on drivers/counter/Kconfig.gecko:
```diff
@@ -1,17 +1,31 @@
+	bool "Silicon Labs Gecko Time (RTCC) persistence"
+	default n
+	depends on DT_HAS_SILABS_GECKO_RTCC_ENABLED && SOC_FAMILY_EXX32
+	select SOC_GECKO_RMU
+	select SOC_GECKO_CMU
+	select SOC_GECKO_RTCC
```
```suggestion
```

DRuffer-tmo replied:
How else are you suggesting that I enable RMU settings?

fkokosinski commented on drivers/counter/Kconfig.gecko:
```diff
@@ -1,17 +1,31 @@
+	select SOC_GECKO_RTCC
+	help
+	  Enable the Gecko RTCC module to preserve the counter operations
+	  across EM4 resets. This mode will also set the Posix time.
```
```suggestion
	  across EM4 resets. This mode will also set the POSIX time.
```

DRuffer-tmo replied:
Ok

fkokosinski commented on drivers/counter/counter_gecko_rtcc.c:
```diff
@@ -11,9 +12,12 @@
 #include <errno.h>
 #include <zephyr/kernel.h>
 #include <zephyr/device.h>
+#include <zephyr/posix/time.h>
 #include <soc.h>
 #include <em_cmu.h>
+#include <em_rmu.h>
 #include <em_rtcc.h>
+#include <zephyr/sys/timeutil.h>
```
Please ifdef these new includes with `CONFIG_PERSISTENT_GECKO_RTCC`.

fkokosinski commented on drivers/counter/counter_gecko_rtcc.c:
```diff
@@ -219,12 +223,68 @@ static uint32_t counter_gecko_get_pending_int(const struct device *dev)
+	return clock_settime(CLOCK_REALTIME, &tp);
+}
+
+/*! \brief Pin bootloader definitions.
+ *
+ * The UART Bootloader (e.g. First Stage Bootloader) also uses the RTCC so it must
+ * be disabled if we are preserving the time in the RTCC.
+ *
+ * Here is a list of the devices supported by the UART Bootloader:
+ * See: https://www.silabs.com/documents/public/application-notes/an0003-efm32-uart-bootloader.pdf
+ * Note that the CONFIG symbols are artificial when they are terminated with a $.
+ */
+#if    (/* EFM32 Series 0: */ \
+	defined(CONFIG_SOC_SERIES_EFM32G$)	/*!< EFM32 Gecko */ ||	\
+	defined(CONFIG_SOC_SERIES_EFM32GG$)	/*!< EFM32 Giant Gecko */ || \
+	defined(CONFIG_SOC_SERIES_EFM32WG)	/*!< EFM32 Wonder Gecko */ || \
+	defined(CONFIG_SOC_SERIES_EFM32LG$)	/*!< EFM32 Leopard Gecko */ || \
+	defined(CONFIG_SOC_SERIES_EFM32TG$)	/*!< EFM32 Tiny Gecko */ || \
+	defined(CONFIG_SOC_SERIES_EFM32ZG$)	/*!< EFM32 Zero Gecko */ || \
+	defined(CONFIG_SOC_SERIES_EFM32HG$)	/*!< EFM32 Happy Gecko */ || \
+	/* EZR32 Series 0: */ \
+	defined(CONFIG_SOC_SERIES_EZR32WG$)	/*!< EZR32 Wonder Gecko */ || \
+	defined(CONFIG_SOC_SERIES_EZR32LG$)	/*!< EZR32 Leopard Gecko */ || \
+	defined(CONFIG_SOC_SERIES_EZR32HG$)	/*!< EZR32 Happy Gecko */ || \
+	/* EFM32 Series 1: */ \
+	defined(CONFIG_SOC_SERIES_EFM32PG1$)	/*!< EFM32 Pearl Gecko (Rev C onwards) */ || \
+	defined(CONFIG_SOC_SERIES_EFM32PG12B)	/*!< EFM32 Pearl Gecko (Rev C onwards) */ || \
+	defined(CONFIG_SOC_SERIES_EFM32JG1$)	/*!< EFM32 Jade Gecko (Rev C onwards) */ || \
+	defined(CONFIG_SOC_SERIES_EFM32JG12$)	/*!< EFM32 Jade Gecko (Rev C onwards) */ || \
+	defined(CONFIG_SOC_SERIES_EFM32GG11$)	/*!< EFM32 Giant Gecko GG11 */ || \
+	defined(CONFIG_SOC_SERIES_EFM32GG12$)	/*!< EFM32 Giant Gecko GG12 */ || \
+	defined(CONFIG_SOC_SERIES_EFM32TG11$)	/*!< EFM32 Tiny Gecko 11 */)
+#define LB_CLW0           (*((volatile uint32_t *)(LOCKBITS_BASE)+122))
+#define LB_CLW0_PINBOOTLOADER    (1 << 1)
+#endif /* EFM32, EZR32 SOC Series */
+#endif /* PERSISTENT_GECKO_RTCC */
```
IMO this belongs in the SoC specific startup code, not in the driver itself.

DRuffer-tmo replied:
In zephyrproject/zephyr/soc/arm/silabs_exx32/common/soc.c?

fkokosinski commented on drivers/counter/counter_gecko_rtcc.c:
```diff
@@ -279,6 +339,19 @@ static int counter_gecko_init(const struct device *dev)
 	CMU_ClockSelectSet(cmuClock_LFA, cmuSelect_LFXO);
 #endif
 
+#ifdef LB_CLW0
+	/* Turn off the bootloader enable bit to prevent RTCC changes */
+	LB_CLW0 &= !LB_CLW0_PINBOOTLOADER;
+#endif /* LB_CLW0 */
```
Ditto

fkokosinski commented on drivers/counter/counter_gecko_rtcc.c:
```diff
@@ -279,6 +339,19 @@ static int counter_gecko_init(const struct device *dev)
+	LB_CLW0 &= !LB_CLW0_PINBOOTLOADER;
+#endif /* LB_CLW0 */
+
+#ifdef CONFIG_PERSISTENT_GECKO_RTCC
+	/* Turn on reset limited mode to preserve RTCC counts */
+	RMU_ResetControl(rmuResetWdog, rmuResetModeLimited);
+	RMU_ResetControl(rmuResetCoreLockup, rmuResetModeLimited);
+	RMU_ResetControl(rmuResetSys, rmuResetModeLimited);
+	RMU_ResetControl(rmuResetPin, rmuResetModeLimited);
+#endif /* CONFIG_PERSISTENT_GECKO_RTCC */
```
Again - please move it to SoC/board platform code.

fkokosinski commented on boards/arm/efm32pg_stk3402a/Kconfig.defconfig:
```diff
@@ -16,4 +18,9 @@ config CMU_HFXO_FREQ
 config CMU_LFXO_FREQ
 	default 32768
 
+config COUNTER_GECKO_RTCC
+	bool
+	default y
+	depends on COUNTER
```
Why is this needed? `COUNTER_GECKO_RTCC` should be enabled by the `DT_HAS_*_ENABLED` macro.

DRuffer-tmo replied:
I'm sorry, but I am not aware of the DT_HAS_*_ENABLED macro. Can you give me a link to where it is defined?

DRuffer-tmo replied:
I see it's usage below and will see if I can fix this properly

fkokosinski commented:
Also - thanks a lot for separating out these changes! It makes it a lot easier to review :)

DRuffer-tmo commented on CODEOWNERS:
```diff
@@ -114,7 +114,7 @@
 /boards/arm/cy8ckit_062_wifi_bt/          @ifyall @npal-cy
 /boards/arm/cy8cproto_062_4343w/          @ifyall @npal-cy
 /boards/arm/disco_l475_iot1/              @erwango
-/boards/arm/efm32pg_stk3401a/             @rdmeneze
```
Ah, I see what you are saying now. My mistake!

DRuffer-tmo commented on boards/arm/efm32pg_stk3402a/efm32pg_stk3402a_defconfig:
```diff
@@ -7,5 +9,10 @@ CONFIG_CONSOLE=y
 CONFIG_CMU_HFCLK_HFXO=y
+CONFIG_PERSISTENT_GECKO_RTCC=y
+CONFIG_NOTIFY=y
+CONFIG_REBOOT=y
```
At the moment, I don't remember why I needed CONFIG_NOTIFY, so I'll just remove it for now, but CONFIG_REBOOT provides a reliable way to test the RTCC persistence in the shell module, so I will put that into a new samples/subsys/shell/shell_module/boards/ folder. Is that ok?

DRuffer-tmo commented on drivers/counter/Kconfig.gecko:
```diff
@@ -1,17 +1,31 @@
+
+config PERSISTENT_GECKO_RTCC
+	bool "Silicon Labs Gecko Time (RTCC) persistence"
+	default n
```
Ok

DRuffer-tmo commented on drivers/counter/Kconfig.gecko:
```diff
@@ -1,17 +1,31 @@
 	bool "Silicon Labs Gecko Counter (RTCC) driver"
 	default y
-	depends on DT_HAS_SILABS_GECKO_RTCC_ENABLED
+	depends on DT_HAS_SILABS_GECKO_RTCC_ENABLED && SOC_FAMILY_EXX32
```
In both cases, I assume?

DRuffer-tmo committed:
boards: arm: Add CONFIG_PERSISTENT_GECKO_RTCC to efm32pg_stk3402a

Correct mistake in CODEOWNERS.
Remove CONFIG_NOTIFY and move CONFIG_REBOOT to
samples/subsys/shell/shell_module/boards/efm32pg_stk3402a.config.
Initial changes to drivers/counter/Kconfig.gecko.
Conditionalize added include files in counter_gecko_rtcc.c.
Replace CONFIG_COUNTER_GECKO_RTCC with CONFIG_PERSISTENT_GECKO_RTCC in
subsys/shell/modules/date_service.c.

Signed-off-by: Dennis Ruffer <Dennis.Ruffer1@T-Mobile.com>

DRuffer-tmo committed:
boards: arm: Add CONFIG_PERSISTENT_GECKO_RTCC to efm32pg_stk3402a

Move PINBOOTLOADER and RMU settings to SoC/board platform code.

Signed-off-by: Dennis Ruffer <Dennis.Ruffer1@T-Mobile.com>
